Add support for rendering to slices of 3D texture views and single layered 2D-Array texture views

teoxoy opened the pull request:
**Connections**
Resolves https://github.com/gfx-rs/wgpu/issues/7313
Resolves https://github.com/gfx-rs/wgpu/issues/6040
Resolves https://github.com/gfx-rs/wgpu/issues/4764

**Description**

- allows 2D-Array texture views with a single layer to be used as render attachments
- adds support for rendering to slices of 3D texture views

Note:
- ~Vulkan implementation is not part of this PR, I will open a follow-up for that since it's more complicated due to the backend caching framebuffers.~
- We could clear 3D textures via render passes as well, I will look into it for a follow-up.

**Testing**
Added tests.

**Squash or Rebase?**
Rebase

teoxoy committed:
[wgpu-core] allow 2D-Array texture views with a single layer to be used as render attachments

teoxoy committed:
add `RenderPassColorAttachment::depth_slice`

teoxoy committed:
[wgpu-hal] make `begin_render_pass` fallible

teoxoy committed:
[d3d12,metal,gl] add support for rendering to slices of 3D textures

ErichDonGubler commented on wgpu/src/api/render_pass.rs:
```diff
@@ -522,6 +522,8 @@ static_assertions::assert_impl_all!(RenderPassTimestampWrites<'_>: Send, Sync);
 pub struct RenderPassColorAttachment<'tex> {
     /// The view to use as an attachment.
     pub view: &'tex TextureView,
+    /// The depth slice index of a 3D view. It must not be proviced if the view is not 3D.
```
typo: s/proviced/provided

```suggestion
    /// The depth slice index of a 3D view. It must not be provided if the view is not 3D.
```

ErichDonGubler commented on wgpu-hal/src/gles/queue.rs:
```diff
@@ -126,13 +127,18 @@ impl super::Queue {
                     };
                 } else if is_layered_target(target) {
+                    let layer = if target == glow::TEXTURE_3D {
+                        depth_slice.unwrap() as i32
```
question(non-blocking): Looking around this code, it seems that panicking is the norm when there are backend errors. 🤔 This feels…wrong? Like, shouldn't we be using internal error machinery instead?

thought: An `expect` here might be nice, but IDK if that's worth it over the existing `unwrap` diagnostic.

teoxoy replied:
> Looking around this code, it seems that panicking is the norm when there are backend errors. 🤔 This feels…wrong? Like, shouldn't we be using internal error machinery instead?

When you say "backend errors" do you mean errors from the native API or situations like here where the backend expects there is a `depth_slice` provided if the view is a 3D view? If we unwrap on errors originating from the native API, we shouldn't do that. In cases like this one, where only wgpu-core could have screwed up (by not validating) I think it's fine.

> An `expect` here might be nice, but IDK if that's worth it over the existing `unwrap` diagnostic.

I think the `unwrap` is fine, there is enough context around the `unwrap` to figure out what went wrong.

ErichDonGubler commented on wgpu-core/src/device/resource.rs:
```diff
@@ -1421,8 +1428,8 @@ impl Device {
-                    && resolved_dimension == TextureViewDimension::D2Array))
+                || resolved_dimension == TextureViewDimension::D2Array
+                || resolved_dimension == TextureViewDimension::D3)
             {
```
suggestion: It might be nice to use [`slice::contains`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.contains) instead here.

ErichDonGubler commented on wgpu-core/src/device/resource.rs:
```diff
@@ -1124,17 +1135,13 @@ impl Device {
 
         let clear_mode = if hal_usage
             .intersects(wgt::TextureUses::DEPTH_STENCIL_WRITE | wgt::TextureUses::COLOR_TARGET)
+            && desc.dimension == wgt::TextureDimension::D2
```
question: What's the follow-up issue for clearing 3D textures? 👀 Also, I'm not sure that I understand the consequences of not having a `clear_mode`, but isn't it necessary for having consistently initialized textures across platforms?

ErichDonGubler replied:
question: Shouldn't we still do the same thing for `D1` textures? Or...oh, I bet we'll just run into validation down below if that's the case, eh?

teoxoy replied:
> Also, I'm not sure that I understand the consequences of not having a `clear_mode`, but isn't it necessary for having consistently initialized textures across platforms?

It will be set to `BufferCopy` which is what we currently use as well. Previously only 2D textures could have `TextureUses::DEPTH_STENCIL_WRITE` or `TextureUses::COLOR_TARGET` usages, so only 2D textures took this path before.

> Shouldn't we still do the same thing for D1 textures? Or...oh, I bet we'll just run into validation down below if that's the case, eh?

I removed the match for 1D textures since they were never taking that path anyway (they can't be render attachments).

teoxoy replied:
> What's the follow-up issue for clearing 3D textures?

There isn't one yet, I'm working on it and might get it into this PR, or open a new PR that depends on this one. Let me know what you'd prefer.

ErichDonGubler replied:
> There isn't one yet, I'm working on it and might get it into this PR.

A race, then, between adding more to the PR and filing an issue if it's merged before, eh? 😁

teoxoy replied:
(I ended up editing the comment I guess while you were answering since I realized it might not be nice to add more onto the reviewer)

ErichDonGubler commented on wgpu-hal/src/dx12/device.rs:
```diff
@@ -725,6 +732,8 @@ impl crate::Device for super::Device {
             device: self.raw.clone(),
             shared: Arc::clone(&self.shared),
             mem_allocator: self.mem_allocator.clone(),
+            rtv_pool: Arc::clone(&self.rtv_pool),
```
thought: `Arc::clone` is the same as calling `expr.clone()`, unlike `Rc::clone`. I wonder why we call it this way here...but maybe it's to avoid `clone`-ing something expensive in the case of a refactor?

teoxoy replied:
I think I've mostly done it to be consistent with the code around, we use `Arc::clone` in a bunch of places.

ErichDonGubler reviewed (commented):
Abandoning a full review for now because I just discovered something urgent to do, here's what I've got so far.

crowlKats reviewed (commented):
Deno part LGTM

teoxoy committed:
[vk] require `VK_KHR_maintenance1` (in preparation for creating 2D texture views from slices of 3D textures)

teoxoy committed:
[vk] remove usage of imageless framebuffers

Imageless framebuffers are not needed, we know which views will be attached to the render pass already. Even if we wanted to more aggressively cache imageless framebuffers that wouldn't be possible since they still require specifiying view usage, texture usage and view formats.

Removing usage of imageless framebuffers simplifies the code substantially.

teoxoy committed:
[vk] fix `raw_pass` not being part of `FramebufferKey`

teoxoy committed:
[vk] move framebuffer cache into `CommandEncoder`

This makes texture view destruction cheap since users of wgpu-hal are required to keep resources referenced by encoders alive for at least as long as the encoder is alive.

This is also a prerequisite to implement rendering to slices of 3D textures (by creating temporary texture views).

On the other hand the cache won't be as effective but that is probably ok, we can reevaluate the implementation if it turns out to be a problem.

teoxoy committed:
[vk] add `TextureView::raw_format`

teoxoy committed:
[vk] add support for rendering to slices of 3D textures

teoxoy commented:
I ended up finishing the Vulkan implementation and pushed it.

ErichDonGubler commented on wgpu-core/src/command/render.rs:
```diff
@@ -1126,6 +1140,71 @@ impl<'d> RenderPassInfo<'d> {
+                ));
+            }
+
+            fn check_attachment_overlap(
+                attachment_set: &mut crate::FastHashSet<(crate::track::TrackerIndex, u32, u32)>,
+                view: &TextureView,
+                depth_slice: Option<u32>,
+            ) -> Result<(), ColorAttachmentError> {
+                let mut insert = |slice| {
+                    let mip_level = view.desc.range.base_mip_level;
+                    if attachment_set.insert((view.tracking_data.tracker_index(), mip_level, slice))
+                    {
+                        Ok(())
+                    } else {
+                        Err(ColorAttachmentError::SubresourceOverlap {
+                            view: view.error_ident(),
+                            mip_level,
+                            depth_or_array_layer: slice,
+                        })
+                    }
+                };
+                match view.desc.dimension {
+                    TextureViewDimension::D2 => {
+                        insert(view.desc.range.base_array_layer)?;
+                    }
+                    TextureViewDimension::D2Array => {
+                        for layer in view.selector.layers.clone() {
+                            insert(layer)?;
+                        }
+                    }
+                    TextureViewDimension::D3 => {
+                        insert(depth_slice.unwrap())?;
+                    }
+                    _ => unreachable!(),
+                };
+                Ok(())
+            }
+
+            check_attachment_overlap(&mut attachment_set, color_view, at.depth_slice)?;
+
```
nitpick: With the commit titled `[d3d12,metal,gl] add support for rendering to slices of 3D textures`, `check_attachment_overlap` seems like overlap validation for color attachments was missing, even beyond 3D textures. This seems worth having added before 3D textures, to distinguish the support for 3D + depth slices from the validation missing overall?

Definitely don't go back and fix this now, but that might have made review a bit easier.

ErichDonGubler commented on wgpu-hal/src/dx12/mod.rs:
```diff
@@ -918,8 +921,10 @@ impl Texture {
     raw_format: Dxgi::Common::DXGI_FORMAT,
     aspects: crate::FormatAspects,
-    /// only used by resolve
-    target_base: (Direct3D12::ID3D12Resource, u32),
```
note: Ah, the second member of `TextureView::target_base` was the `subresource_index`, gotcha.

ErichDonGubler commented on tests/tests/wgpu-gpu/render_target.rs:
```diff
@@ -233,3 +233,188 @@ async fn run_test(
+    const fn size_for_mips(mips: u32) -> u64 {
+        let mut out: u64 = 0;
+        let mut mip = 0;
+        while mip < mips {
```
thought: Wow, that's annoying that we can't do `for` loops on `Range`s yet in `const fn`s.

ErichDonGubler commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -943,13 +943,8 @@ impl PhysicalDeviceProperties {
-                // `VK_AMD_negative_viewport_height` is obsoleted by `VK_KHR_maintenance1` and must not be enabled alongside it
-                extensions.push(amd::negative_viewport_height::NAME);
-            }
+            // Require `VK_KHR_maintenance1`
+            extensions.push(khr::maintenance1::NAME);
```
question: Should this require a `CHANGELOG` entry? I imagine we fell back to `VK_AMD_negative_viewport_height` for a reason.

ErichDonGubler replied:
issue: Ah, actually, this _whole thing_ is missing a `CHANGELOG` entry and is a pretty big deal, too. 😅

Trusting that this will be handled, despite my having approved already.

teoxoy replied:
It's mostly old drivers that don't support `VK_KHR_maintenance1` but support `VK_AMD_negative_viewport_height` https://github.com/gpuweb/gpuweb/issues/4251#issuecomment-1656125294 - but I can note it in the changelog entry.

ErichDonGubler commented on wgpu-hal/src/vulkan/adapter.rs:
nitpick: For the commit `[vk] remove usage of imageless framebuffers`, there's a typo in the commit message: s/Remving/Removing

ErichDonGubler commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -44,9 +44,6 @@ pub struct PhysicalDeviceFeatures {
     pub(super) descriptor_indexing:
         Option<vk::PhysicalDeviceDescriptorIndexingFeaturesEXT<'static>>,
 
-    /// Features provided by `VK_KHR_imageless_framebuffer`, promoted to Vulkan 1.2.
-    imageless_framebuffer: Option<vk::PhysicalDeviceImagelessFramebufferFeaturesKHR<'static>>,
-
```
praise: Nice on the imageless framebuffers!

ErichDonGubler commented on wgpu-hal/src/vulkan/command.rs:
```diff
@@ -52,6 +51,26 @@ impl super::CommandEncoder {
+        Ok(match self.framebuffers.entry(key) {
+            Entry::Occupied(e) => *e.get(),
+            Entry::Vacant(e) => {
+                let vk_info = vk::FramebufferCreateInfo::default()
+                    .render_pass(e.key().raw_pass)
+                    .width(e.key().extent.width)
+                    .height(e.key().extent.height)
+                    .layers(e.key().extent.depth_or_array_layers)
+                    .attachments(&e.key().attachments);
```
question: Would destructuring be helpful here for refactor-proof code?

teoxoy replied:
That's a good point since all the fields of the key should be used, if not they should be removed; will update.

ErichDonGubler commented on wgpu-hal/src/vulkan/mod.rs:
```diff
@@ -602,13 +602,6 @@ struct RenderPassKey {
-    attachments: ArrayVec<vk::ImageView, { MAX_TOTAL_ATTACHMENTS }>,
-    extent: wgt::Extent3d,
-}
-
```
question(non-blocking): Was there a reason you moved `FramebufferKey`?

teoxoy replied:
I moved it closer to where it will be used (in the command encoder).

ErichDonGubler commented on wgpu-hal/src/vulkan/device.rs:
```diff
@@ -1260,11 +1260,12 @@ impl crate::Device for super::Device {
         desc: &crate::TextureViewDescriptor,
     ) -> Result<super::TextureView, crate::DeviceError> {
         let subresource_range = conv::map_subresource_range(&desc.range, texture.format);
+        let raw_format = self.shared.private_caps.map_texture_format(desc.format);
```
note: It took me a bit to understand this change with keeping a `raw_format`, but for myself: Computing the raw format with `private_caps.map_texture_format` already happened once, and we're just trading some space for avoiding the lookup again later.

ErichDonGubler commented on wgpu-hal/src/vulkan/command.rs:
```diff
@@ -747,6 +772,18 @@ impl crate::CommandEncoder for super::CommandEncoder {
 
         for cat in desc.color_attachments {
             if let Some(cat) = cat.as_ref() {
+                let color_view = if cat.target.view.dimension == wgt::TextureViewDimension::D3 {
+                    let key = super::TempTextureViewKey {
+                        texture: cat.target.view.raw_texture,
+                        format: cat.target.view.raw_format,
+                        mip_level: cat.target.view.base_mip_level,
+                        depth_slice: cat.depth_slice.unwrap(),
+                    };
+                    self.make_temp_texture_view(key)?
+                } else {
+                    cat.target.view.raw
+                };
+
```
question: Is destructuring here interesting for more refactor-proof code?

teoxoy replied:
I don't think so as we don't use the view's `layers` of `format` and if we add other fields in the future they most likely don't have to be used here either.

ErichDonGubler reviewed (approved):
LGTM! I have a bunch of minor stuff I wrote, but overall: 💯

ErichDonGubler commented on wgpu-hal/src/vulkan/command.rs:
```diff
@@ -52,6 +51,51 @@ impl super::CommandEncoder {
+                    .attachments(&e.key().attachments);
+
+                let raw = unsafe { self.device.raw.create_framebuffer(&vk_info, None).unwrap() };
+                *e.insert(raw)
```
thought: Sigh, I wish nice helpers like `Entry::or_insert_with` didn't have to be sacrificed when we need to bail out with `?` if something goes wrong.

teoxoy committed:
[vk] use destructuring for cache keys

If fields are unused, they shouldn't be part of the cache key.

teoxoy committed:
add changelog entry
